### DISCUSSION TITLE
GSoC 2026 – Interest in Behavioral Evaluation Test Framework (Idea #2)

### DISCUSSION THREAD

**Discussion by KenWuqianghao** [google-gemini/gemini-cli#20482] [Ideas]
Hi, I'm Ken Wu, a final-year CS student at the University of Waterloo. I've been building LLM eval systems for the past year: at Nokia I built an eval framework for fine-tuned Qwen models that caught overfitting we couldn't see from loss curves alone, and at August I built a round-robin scoring framework for 15+ LLM agents with LLM-as-Judge. I've been reading through the gemini-cli eval code for the past week and wanted to share what I found before drafting a proposal.

cc @gundermanc @srithreepo

---

### What already exists

The `evals/` directory is more capable than the issue description (#18257) makes it sound. A few things I noticed:

`evalTest` in `evals/test-helper.ts` has a two-tier policy system: `ALWAYS_PASSES` runs in CI on every PR, `USUALLY_PASSES` gets tracked nightly. So the framework already handles LLM non-determinism instead of treating every flaky result as a test bug. That's good.

`TestRig` (`packages/test-utils/src/test-rig.ts`) does a lot more than I expected. It sets up isolated CLI environments with custom file systems, runs prompts, waits for specific tool calls (you can even validate arguments via callback), and gives you the full telemetry logs afterward through `readToolLogs()`. It also handles git init, node_modules symlinking, agent acknowledgment files, activity log capture.

The nightly pipeline (`evals-nightly.yml`) runs everything 3 times across 6 models, aggregates pass rates, and keeps 10-run history. There's even `/fix-behavioral-eval` that auto-investigates regressions by pulling nightly results with `gh` CLI and suggesting prompt fixes. So the full loop already works: write eval, run nightly, detect regression, auto-fix.

### Where the gaps are

There are ~20 eval files right now. Comparing that against the GSoC target of 50+ across debugging, refactoring, new features, and code review, a few things stand out:

**No debugging evals.** Nothing hands the agent a broken codebase and checks if it can find and fix the bug. `validation_fidelity.eval.ts` tests refactoring but always starts from working code.

**No code review evals.** Can the agent spot a security issue in a diff? Can it flag a breaking API change? No idea, it's never tested.

**No multi-file reasoning.** Most evals use 1-3 files. Real tasks usually need the agent to figure out which files are even relevant before it can start working.

**No efficiency tracking.** The nightly pipeline knows if the agent passed or failed, but not how it got there. This was a big lesson from my time at August: one agent solves a task in 3 tool calls, another thrashes around for 30 and still technically passes. `readToolLogs()` already captures this data, nobody's aggregating it.

**No taxonomy.** Evals are grouped by feature (grep, memory, plan mode). If you want to know "how good is the agent at debugging?" there's no way to slice the data that way.

### What I'd build

**1. Task taxonomy**

I'd extend `EvalCase` with category metadata rather than replacing anything:

```typescript
interface EvalScenario extends EvalCase {
  category: 'debugging' | 'refactoring' | 'new-feature' | 'code-review' | 'exploration';
  complexity: 'single-file' | 'multi-file' | 'cross-package';
  expected_tools: string[];
}

interface EvalScore {
  passed: boolean;
  tool_calls_total: number;
  redundant_reads: number;
  wasted_explorations: number;
  self_corrections: number;
}
```

`evalTest` and the policy system stay the same. This just adds fields so the nightly report can group results by category.

**2. New scenarios**

Debugging:
- Off-by-one bug with a failing test. Agent should fix `< length` to `<= length` without touching unrelated code.
- Broken import path in a TypeScript project. Should fix the import, not rewrite the module.
- Null reference buried three function calls deep. Can the agent actually trace the stack?

Code review:
- Diff that introduces unsanitized user input in a shell command. Agent should flag it.
- Refactor that silently changes public API behavior. Agent should catch the break.

Multi-file:
- Rename a type used across 5+ files, verify all references updated and type checker passes. Similar to the existing `validation_fidelity.eval.ts` but bigger.
- New feature that touches a type definition, a service, and a test file. Order matters.

**3. Efficiency scoring**

This is the part I care about most, because I ran into exactly this problem at August. Pass/fail hides huge quality differences. I'd pull metrics from `readToolLogs()` after each run:

- Useful tool calls vs total calls
- Files read but never referenced in the solution
- Whether the agent recovered after going down a wrong path
- Same tool called with the same arguments multiple times

None of this would block CI. It'd show up in the nightly report as trend lines alongside pass rates.

**4. Better reports**

Extend `scripts/aggregate_evals.js` to break results down by category instead of only by test name. Add regression alerts when a category's pass rate drops below its rolling average.

**5. `gemini --eval` CLI command (stretch goal)**

This was [suggested on #18257](https://github.com/google-gemini/gemini-cli/issues/18257#issuecomment-3883063320) -- letting users run custom eval suites from a JSON file, like how [ADK's eval command](https://google.github.io/adk-docs/evaluate/#how-to-run-evaluation-with-the-adk) works. Would be useful for extension authors. I'd treat this as a stretch goal depending on how the first four pieces go.

### Questions for mentors

1. Should I prioritize getting to 50+ scenarios first, or is the efficiency scoring system more useful to you right now?

2. Is there historical nightly eval data somewhere I can use for baselines? The aggregate script references past runs but artifacts only seem to last 7 days.

3. The nightly pipeline tests 6 models. Should the framework track per-model baselines, or is one "primary model" baseline enough?

4. How much priority should `gemini --eval custom.json` get within the 175 hours?

I'm working on a PR that adds a debugging-category eval as a first contribution. Let me know if there's a different area that'd be more useful.

| | |
|---|---|
| Portfolio | [kenwu.is-a.dev](https://kenwu.is-a.dev/) |
| GitHub | [@KenWuqianghao](https://github.com/KenWuqianghao) |
| LinkedIn | [kenwuu](https://linkedin.com/in/kenwuu) |
| Resume | [Google Drive](https://drive.google.com/file/d/11TiGQ-JxqmLQ-TJ24Jui8V9kXsI6QZld/view) |

**Comment by ATREAY** [comment]
Hi everyone, I’m Atreay, a first-year Master's CSE student at IIT Hyderabad. My background is mainly in AI/ML and developer tooling, and I’ve recently started exploring the Gemini-CLI repository while preparing for GSoC.

Ken’s breakdown of the current eval infrastructure (especially around TestRig and the nightly evaluation pipeline) was really helpful for understanding how the existing setup works.

While looking through the `evals` directory and the discussion here, I was wondering whether future evaluation scenarios might benefit from being structured with different difficulty levels (for example simple bug fixes, multi-file refactoring tasks, and debugging tasks using small repos with failing tests). That might make it easier to track regressions across different task types.

I also noticed that tool logs are available via `readToolLogs()`. Would it be useful to track additional metrics like tool-usage efficiency or repeated file reads alongside the current pass/fail evaluation?

I’m still exploring the eval framework and would appreciate any pointers on good starting points for contributors interested in expanding the evaluation scenarios.

**Comment by manas-raj999** [comment]
Hi Ken, really insightful breakdown of the current eval infrastructure ,especially the parts about `TestRig` and the CI policy tiers. It clarified a lot about how the framework already handles non-determinism.

One thought I had while reading your proposal: since tool logs already capture detailed execution traces, wouldn't introducing efficiency metrics like redundant reads or tool call counts risk combining exploration behavior with inefficiency?

For complex debugging or multi-file reasoning tasks, an agent may had to explore several files before finally deciding  on the correct fix. If efficiency metrics are introduced too early, they might unintentionally bias evaluations toward shorter execution paths rather than better reasoning strategies or solutions.

Maybe it would be interesting to first expand the scenario diversity (debugging, code review, multi-file repos) and then derive efficiency baselines from that richer dataset.

Curious to hear your perspective - I'm currently exploring the architecture and framework and was just in need for an insight like this , I would definitely try to understand where new evaluation signals would fit best next.

**Comment by aniruddhaadak80** [comment]
From my point of view, the most valuable part of this idea is the taxonomy and scenario coverage first, then efficiency scoring once there is enough diversity to interpret the metrics correctly. If efficiency is introduced too early, it can punish legitimate exploration on harder tasks. A category level view across debugging, code review, and multi file work would probably give maintainers clearer signals sooner.
